### PR TITLE
sql: remove leftovers of interleaved tables

### LINKED ISSUE #42947: sql: Remove interstices allocation from `SpanBuilder`
Once #42688 lands, there are some optimizations that can be made. One is to remove the large upfront allocation of interstices that the span_builder performs. This buffer is used by the `constraint.Spans` conversion into a `roachpb.Span` by the optimizer. If those methods can be changed to not need the buffer, then a large allocation can be avoided.

### PR DESCRIPTION
**Pull request by yuzefovich**
This commit removes most of the remaining mentions of the interleaved
tables in the code base.

Addresses: #69150.
Fixes: #42947.

Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
This PR touches a variety of packages, but I picked @mgartner and @rharding6373 as the main reviewers since most changes are in the SQL land. @postamar maybe you'd be curious to take a look here too.

File: pkg/ccl/backupccl/show.go [hunk 1]
```diff
@@ -362,11 +362,11 @@ func backupShowerDefault(
...
 				descSizes := make(map[descpb.ID]RowCount)
 				for _, file := range manifest.Files {
-					// TODO(dan): This assumes each file in the backup only contains
-					// data from a single table, which is usually but not always
-					// correct. It does not account for interleaved tables or if a
-					// BACKUP happened to catch a newly created table that hadn't yet
-					// been split into its own range.
+					// TODO(dan): This assumes each file in the backup only
+					// contains data from a single table, which is usually but
+					// not always correct. It does not account for ia BACKUP
```
Comment (postamar) [inline pkg/ccl/backupccl/show.go:367]: typo

Comment (postamar) [inline pkg/ccl/backupccl/show.go:367]: typo

**Review by postamar — APPROVED** [review]
Very nice, thank you for doing this! Deleting code feels so good. If there are any traces of code related to interleaves left after this change, I feel that they really can only be quite small and inconsequential. It seems you've been quite thorough.

I can't comment on the deeper sql-queries-y code but for the rest, which I mostly understand, your change looks fine and much like what I'd expect. LGTM.

**Comment by msbutler** [comment]
the comment change in `backupccl` looks fine, once you fix the typo! This was the only place I looked closely.

**Review by msbutler — APPROVED** [review]

**Comment by msbutler** [comment]
@yuzefovich it might be worth adding a cdc reviewer for the changefeed changes.

**Review by yuzefovich — COMMENTED** [review]
TFTRs!

@miretskiy could you please take a look at CDC changes?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72497#-:-Mo-JgY29_HJB2ntZarK:bpkb1co)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @postamar, and @rharding6373)

___
*[pkg/ccl/backupccl/show.go, line 367 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/72497#-MnmizHo4xH_GKs7GwPf-r1-367:-Mo-JXqq9JASKNqZL0vV:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a4645d3ea15159f3f7a30166ffd77f4b1c1e632/pkg/ccl/backupccl/show.go#L367)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

typo
</blockquote></details>

Done.

**Review by rharding6373 — APPROVED** [review]
Nice! :lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72497#-:-Mo0oCe66GUw3eX0MUu6:bfd61qr)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @postamar, and @yuzefovich)

___
*[pkg/sql/opt/exec/execbuilder/testdata/distsql_tighten_spans, line 183 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72497#-Mo0djtF3UpZwEv9d3RN:-Mo0djtF3UpZwEv9d3RO:b34sx3m) ([raw file](https://github.com/cockroachdb/cockroach/blob/12d0fb3ce288812cb80c546df5ddb1957f48e912/pkg/sql/opt/exec/execbuilder/testdata/distsql_tighten_spans#L183)):*
> ```
>   spans: /1-/1/9223372036854775807/PrefixEnd
> 
> # Table c1 (interleaved table)
> ```

Is this still an interleaved table?
___
*[pkg/sql/opt/exec/execbuilder/testdata/distsql_tighten_spans, line 198 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72497#-Mo0jjtg3NXR8tTDF0le:-Mo0jjtg3NXR8tTDF0lf:b-1vigaa) ([raw file](https://github.com/cockroachdb/cockroach/blob/12d0fb3ce288812cb80c546df5ddb1957f48e912/pkg/sql/opt/exec/execbuilder/testdata/distsql_tighten_spans#L198)):*
> ```
>       spans: -/3/4
> 
> # Table p2 with interleaved index.
> ```

Is this still an interleaved index?

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72497)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @postamar, and @rharding6373)

___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 704 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72497#-Mo1FCzmDT-td8Fsqnou:-Mo1FCzmDT-td8Fsqnov:bkcquv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9baa5da82f060e200dfb6e3d035ebab58eae81d9/pkg/ccl/changefeedccl/changefeed_processors.go#L704)):*
> ```Go
> 	}
> 
> 	if _, ok := c.details.Targets[desc.GetID()]; !ok {
> ```

cc @cockroachdb/cdc 
___
*[pkg/sql/opt/exec/execbuilder/testdata/distsql_tighten_spans, line 183 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72497#-Mo0djtF3UpZwEv9d3RN:-Mo1EufV0r0VazaRTNXl:bg1hvmy) ([raw file](https://github.com/cockroachdb/cockroach/blob/12d0fb3ce288812cb80c546df5ddb1957f48e912/pkg/sql/opt/exec/execbuilder/testdata/distsql_tighten_spans#L183)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Is this still an interleaved table?
</blockquote></details>

No, I wasn't sure what to do about these in the comments. I removed it.
___
*[pkg/sql/opt/exec/execbuilder/testdata/distsql_tighten_spans, line 198 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/72497#-Mo0jjtg3NXR8tTDF0le:-Mo1EzbJ1noAZGm_kPIj:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/12d0fb3ce288812cb80c546df5ddb1957f48e912/pkg/sql/opt/exec/execbuilder/testdata/distsql_tighten_spans#L198)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Is this still an interleaved index?
</blockquote></details>

Removed.
___
*[pkg/storage/mvcc.go, line 3966 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/72497#-Mo1FfZ67Mwh31ILM2Xk:-Mo1FfZ67Mwh31ILM2Xl:b-endn22) ([raw file](https://github.com/cockroachdb/cockroach/blob/d845d68075b8513c1dabeb2b740d5f846014ec8c/pkg/storage/mvcc.go#L3966)):*
> ```Go
> 			}
> 			// Allow a split key before other rows in the same table.
> 			minSplitKey = firstRowKey.PrefixEnd()
> ```

cc @cockroachdb/storage

**Review by HonoreDB — COMMENTED** [review]
CDC changes :lgtm: 

> Reviewed 2 of 33 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72497#-:-Mo4OZSw5ENmeXkI7hco:b-4gjato)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @postamar, and @rharding6373)

**Review by itsbilal — COMMENTED** [review]
Storage changes :LGTM:

> Reviewed 2 of 33 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72497#-:-Mo4dNQM9rlWqpkNMVn3:b-quyhfx)* status: :shipit: complete! 2 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner, @postamar, and @rharding6373)

**Review by mgartner — APPROVED** [review]
:lgtm: So so satisfying.

> Reviewed 25 of 33 files at r1, 1 of 1 files at r2, 7 of 8 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72497#-:-Mo4gR8s16OW5v40xx4g:blnomdc)* status: :shipit: complete! 3 of 0 LGTMs obtained (and 1 stale) (waiting on @HonoreDB, @itsbilal, @mgartner, @postamar, and @rharding6373)

**Comment by yuzefovich** [comment]
Thanks for the reviews everyone!

bors r+

**Comment by yuzefovich** [comment]
Hm, seems like a flake, let's try one more time.

bors r+

**Pull request merged**